overlord/devicestate/users.go

```diff
@@ -0,0 +1,339 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+/*
+ * Copyright (C) 2022 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package devicestate
+
+import (
+	"encoding/json"
+	"fmt"
+	"os/user"
+	"path/filepath"
+	"time"
+
+	"github.com/snapcore/snapd/asserts"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/overlord/assertstate"
+	"github.com/snapcore/snapd/overlord/auth"
+	"github.com/snapcore/snapd/overlord/snapstate"
+	"github.com/snapcore/snapd/overlord/state"
+	"github.com/snapcore/snapd/release"
+	"github.com/snapcore/snapd/strutil"
+)
+
+type UserResponse struct {
+	Username string
+	SSHKeys  []string
+}
+
+var (
+	osutilAddUser = osutil.AddUser
+	osutilDelUser = osutil.DelUser
+	userLookup    = user.Lookup
+)
+
+func RemoveUser(st *state.State, username string) (*auth.UserState, bool, error) {
```

Comment:
These functions need to be documented. The second return value (that in the caller code is referred to as "internal") is especially intriguing. :-) At a first sight, it looks like something that could be part of the error value.

Reply:
TBH I'm not sure what is the right way here.
The motivation for moving `CreateUser` functionality under overlord is to have the ability to call it internally( in case of auto import assertion) as well as calling from daemon as requested over REST API. For later, at least the initial implementation does distinguish between "internal error" and "bad request", while when called internally we treat all errors equally.
So while we can return error already wrapped as "bad request" or "internal error" this would not make sense once called internally. Hence I passed the internal flag, which can be ignored or used to determine the error type.

But I totally agree this is not the bested way to approach it. Happy to change it to something more conventional. I assume the wrapper to the error might be just that.

Reply:
@mardy @MiguelPires I have now introduced wrapper for returned error to have cleaner function signature.
Internal error is something which can be now consumed if needed, otherwise easily ignored. It also made it cleaner when there is no error, as before we still needed to return some "internal" value which had no meaning. So indeed better now.